### PR TITLE
Support DRBD replication

### PR DESCRIPTION
**Pull request by TuxPowered42**

### CODE REVIEW

File: igvm/hypervisor.py [hunk 1]
```diff
@@ -463,21 +464,31 @@ def check_migration(self, vm, hypervisor, offline):
...
+                    self.vm_disk_path(domain.name()),
+                    vm.dataset_obj['disk_size_gib'] * 1024**3,
+                    nc_listener,
+                    tx,
+                )
+            if use_drbd:
+                self.wait_for_sync(vm, target_hypervisor, tx=tx)
+                if vm.is_running():
+                    vm.shutdown(tx=tx)
+                self.stop_drbd(vm, target_hypervisor, tx=tx)
```
Comment (HerrSpace) [inline igvm/hypervisor.py]: Can't you merge the two `if use_drbd` blocks?

**Comment by HerrSpace** [comment]
Please also update the readme. While at it you could drop the whole Development section in there.

File: igvm/drbd.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                dev_size / 512,
+                self.get_device_minor(),
+            )
+        )
+        if tx:
+            tx.on_rollback(
+                # There should be no need for resume because it happens also
+                # via another rollback defined above. Unfortunately it is
+                # needed before DRBD can be shutdown.
+                # WARNING: Potental race between writes to DRBD and underlying
```
Comment (HerrSpace) [inline igvm/drbd.py]: *Potential

File: igvm/drbd.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                # device - potential data loss?
```
Comment (HerrSpace) [inline igvm/drbd.py]: I don't quite get the reason yet. What happens if you don't resume it before?

File: igvm/settings.py [hunk 4]
```diff
@@ -3,6 +3,10 @@
 Copyright (c) 2018, InnoGames GmbH
 """
 
+import logging
+
+logging.getLogger('paramiko').setLevel(logging.WARNING)
+
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/settings.py]: But don't I want exactly that? I added this because for DRBD migration displaying progress bar was impossible because every read of `/proc/drbd` was producing multiple lines of output. Furthermore, I would call this "debug output", even though its level was "info".

File: igvm/host.py [hunk 5]
```diff
@@ -158,6 +159,20 @@ def read_file(self, path):
...
 
+    def put(self, remote_path, local_path, mode='0644'):
+        """ Same as Fabric's put but with working sudo permissions
+
+            Setting permissions on files and using sudo via Fabric's put()
+            seems broken, at least for mounted VM. This is why we run
+            extra commands here.
+        """
+        with self.fabric_settings():
+            tempfile = '/tmp/' + str(uuid1())
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/host.py]: Why easier? How the algorithm works behind the scenes is not important to us.

File: igvm/hypervisor.py (hunk 1 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py]: Maybe even better to have 2 separate functions. Please note that those things have particular order, so no, they can't be easily merged.

File: igvm/drbd.py (hunk 2 above)
Comment (TuxPowered42) [reply igvm/drbd.py]: Fixed.

File: igvm/drbd.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    def build_config(self, peer, tx=None):
+        fd = StringIO(
+            'resource {dev} {{\n'
+            '    net {{\n'
+            '        protocol A;\n'
+            # max-buffers vs MB/s
+            # 4k-150, 8k-233, 12k-330, 16K-397, 24k-561, 32k-700
+            # 32k seems jumpy and might end up at as low aw 250MB/s
+            '        max-buffers 24k;\n'
+            # Buffer sizes don't seem to make difference, at least within one
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/drbd.py]: Fixed.

File: igvm/drbd.py (hunk 3 above)
Comment (TuxPowered42) [reply igvm/drbd.py]: I updated the comment to be more clear about this issue.

File: igvm/drbd.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+    def get_device_minor(self):
+        if self.dev_minor is None:
+            dev_minor = self.hv.run(
+                'stat -L -c "%T" /dev/mapper/{}'
+                .format(self.get_mapper_name()),
+                silent=True,
+            )
+            self.dev_minor = int(dev_minor, 16)
+        return self.dev_minor
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/drbd.py]: I rewrote the code to make device names more user-friendly.

File: igvm/drbd.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+            )
+
+        # Meta device must be zeroed, otherwise DRBD might complain
+        self.hv.run(
+            'dd if=/dev/zero of=/dev/{}/{}_meta bs=1048576 count=256'
+            .format(self.vg_name, self.lv_name)
+        )
+
+    def prepare_lv_override(self, tx=None):
+        """Prepare logical volume do be overriden by DRBD revice"""
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/drbd.py]: Fixed.

File: igvm/drbd.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        self.dev_minor = None
+        self.mapper_name = None
+
+    def get_mapper_name(self):
+        if self.mapper_name is None:
+            self.mapper_name = (
+                '{}/{}'.format(self.vg_name, self.lv_name)
+                .replace('-', '--').replace('/', '-')
+            )
+        return self.mapper_name
```
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/drbd.py]: This is gone from the code.

File: igvm/settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HerrSpace) [reply igvm/settings.py]: You are not changing the level for igvm though, but for every usage of the global paramiko logger anywhere. Daniel made paramiko logging configurable in iggop and you will change our logging configuration inadvertently. The python docs also explicitly ask library devs not to attach loggers, see the Note at the end of https://docs.python.org/2/howto/logging.html#configuring-logging-for-a-library

File: igvm/host.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (HerrSpace) [reply igvm/host.py]: uuid1 requires you to always consider if your code might be called twice so quickly, that one uuid was generated twice as most of uuid1 is a mac and the rest is a timestamp. uuid4 is completely random and if those collide god have mercy on our souls.

File: igvm/settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/settings.py]: Let me check if I can change it just for the progress bar displaying. But no matter if it is possible or not, I consider this to be debug output, not any kind of "info". But well, that is Paramiko, I don't expect much better code quality.

File: igvm/host.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TuxPowered42) [reply igvm/host.py]: In such case we must fix all places where it `uuid1()` is used. Let's do it separately from this DRBD code.

[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/host.py]: Here it is: #12

File: igvm/host.py [hunk 10]
```diff
@@ -146,6 +147,20 @@ def read_file(self, path):
             fabric.api.get(path, fd)
             return fd.getvalue()
 
+    def put(self, remote_path, local_path, mode='0644'):
+        """ Same as Fabric's put but with working sudo permissions
+
+            Setting permissions on files and using sudo via Fabric's put()
+            seems broken, at least for mounted VM. This is why we run
+            extra commands here.
+        """
```
Comment (hasegeli) [inline igvm/host.py]: Could you please not indent your docstrings?

File: igvm/utils/transaction.py [hunk 11]
```diff
@@ -47,11 +48,21 @@ def rollback(self):
...
+                'Checkpoint reached, all previous actions are now permanent'
+            )
+        else:
+            log.debug(
+                'Checkpoint reached, restoring to given position'
+            )
+        self._actions = position
+
+    def copy_checkpoint(self):
+        return copy(self._actions)
```
Comment (hasegeli) [inline igvm/utils/transaction.py]: I am developing a context manager for sub-transactions.

File: igvm/hypervisor.py [hunk 12]
```diff
@@ -674,8 +674,13 @@ def check_netcat(self, port):
                 'Listening netcat already found on destination hypervisor.'
             )
 
-    def kill_netcat(self, port):
-        self.run('pkill -f "^/bin/nc.traditional -l -p {}"'.format(port))
+    def kill_netcat_to_device(self, port):
+        # Warn only because if netcat is already dead, kill will fail.
+        with settings(warn_only=True):
```
Comment (hasegeli) [inline igvm/hypervisor.py]: This wouldn't be necessary once we have sub-transaction support.

File: igvm/vm.py [hunk 13]
```diff
@@ -494,28 +501,23 @@ def run_puppet(self, clear_cert, tx):
...
+            ' --waitforcert=60 --onetime --no-daemonize'
+            ' --skip_tags=chroot_unsafe'
+            ' && touch /tmp/puppet_success'
+            ' | tee {} ;'
+            ' test -f /tmp/puppet_success'
+            .format(
+                self.fqdn,
+                self.dataset_obj['puppet_master'],
+                self.dataset_obj['puppet_ca'],
+                '/var/log/puppetrun_igvm',
```
Comment (hasegeli) [inline igvm/vm.py]: Could you please separate this fix to another commit.

File: igvm/vm.py [hunk 14]
```diff
@@ -494,28 +501,23 @@ def run_puppet(self, clear_cert, tx):
...
             )
+        )
+        if tx:
+            tx.on_rollback('Kill puppet', self.kill_puppet)
```
Comment (hasegeli) [inline igvm/vm.py]: Wasn't it correct to register this before running `puppet`?  We want to kill it if it fails to complete the run.

File: igvm/vm.py [hunk 15]
```diff
@@ -478,6 +478,13 @@ def create_swap(self, size_MiB):
         self.run('/bin/chmod 0600 /swap')
         self.run('/sbin/mkswap /swap')
 
+    def kill_puppet(self):
+        with settings(warn_only=True):
```
Comment (hasegeli) [inline igvm/vm.py]: Do we really need this?

File: igvm/hypervisor.py [hunk 16]
```diff
@@ -463,21 +470,37 @@ def check_migration(self, vm, hypervisor, offline):
                 'configuration (different VLAN).'
             )
 
-    def migrate_vm(self, vm, target_hypervisor, offline, tx):
+    def migrate_vm(self, vm, target_hypervisor, offline, offline_transport, tx):
```
Comment (hasegeli) [inline igvm/hypervisor.py]: This should have a default for backwards compatibility.

File: igvm/hypervisor.py [hunk 17]
```diff
@@ -705,3 +728,22 @@ def device_to_netcat(self, device, size, listener, tx=None):
...
             .format(device, size, *listener)
         )
+
+    def start_drbd(self, vm, peer, tx=None):
+        # Ensure that current domain name is used, this might be non-fqdn
+        # one on source Hypervisor.
+        domain = self._get_domain(vm)
+
+        self.host_drbd = DRBD(self, VG_NAME, domain.name(), vm.fqdn, True, tx)
+        self.peer_drbd = DRBD(peer, VG_NAME, vm.fqdn, vm.fqdn, False, tx)
```
Comment (hasegeli) [inline igvm/hypervisor.py]: It is not nice to save those to the `Hypervisor` object for ever.

File: igvm/hypervisor.py (hunk 12 above)
Comment (hasegeli) [reply igvm/hypervisor.py]: I included the fix as the example of subtransactions: #13

File: igvm/utils/transaction.py (hunk 11 above)
Comment (hasegeli) [reply igvm/utils/transaction.py]: Here it is: #13
